staging/src/k8s.io/kms/encryption/service_test.go

```diff
@@ -361,34 +432,366 @@ func TestDecryptError(t *testing.T) {
+}
+
+func TestRotationRemoteKeyIDChanged(t *testing.T) {
+	ctx := testContext(t)
+
+	var record sync.Map
+
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	remoteKMS := &testRemoteService{keyID: "test-key-id"}
+	localKEKService := newLocalKEKService(ctx, remoteKMS, 10, 5, 1*time.Minute, 100*time.Millisecond, fakeClock)
+	waitUntilReady(t, localKEKService)
+	lk := localKEKService.localKEKTracker.Load()
+	encLocalKEK := lk.encKEK
+
+	var wg sync.WaitGroup
+	for i := 0; i < 3; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resp, err := localKEKService.Encrypt(ctx, "test-uid", []byte("test-plaintext"))
+			if err != nil {
+				t.Fatalf("Encrypt() error = %v", err)
+			}
+			if v, ok := resp.Annotations[referenceKEKAnnotationKey]; !ok || !bytes.Equal(v, encLocalKEK) {
+				t.Fatalf("Encrypt() annotations = %v, want %v", resp.Annotations, encLocalKEK)
+			}
+			record.Store(resp, nil)
+		}()
+	}
+	wg.Wait()
+
+	// check local KEK has only been used 3 times and still under the suggested usage
+	if lk.usage.Load() != 3 {
+		t.Fatalf("local KEK usage = %v, want %v", lk.usage.Load(), 3)
+	}
+
+	fakeClock.Step(30 * time.Second)
+	// change the remote key ID
+	remoteKMS.SetKeyID("test-key-id-2")
+	if _, err := localKEKService.Status(ctx); err != nil {
+		t.Fatalf("Status() error = %v", err)
+	}
+
+	rotated := false
+	// wait for the local KEK to be rotated due to remote key ID change
+	wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+		lk = localKEKService.localKEKTracker.Load()
+		rotated = !bytes.Equal(lk.encKEK, encLocalKEK)
+		return rotated, nil
+	})
+	if !rotated {
+		t.Fatalf("local KEK must have been rotated")
+	}
+
+	// new local KEK must be used for encryption
+	for i := 0; i < 5; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			resp, err := localKEKService.Encrypt(ctx, "test-uid", []byte("test-plaintext"))
+			if err != nil {
+				t.Fatalf("Encrypt() error = %v", err)
+			}
+			if v, ok := resp.Annotations[referenceKEKAnnotationKey]; !ok || !bytes.Equal(v, lk.encKEK) {
+				t.Fatalf("Encrypt() annotations = %v, want %v", resp.Annotations, lk.encKEK)
+			}
+			record.Store(resp, nil)
+		}()
+	}
+	wg.Wait()
+
+	// check we can decrypt data encrypted with the old and new local KEKs
+	record.Range(func(key, _ any) bool {
+		k := key.(*service.EncryptResponse)
+		decryptRequest := &service.DecryptRequest{
+			Ciphertext:  k.Ciphertext,
+			Annotations: k.Annotations,
+			KeyID:       k.KeyID,
+		}
+		if _, err := localKEKService.Decrypt(ctx, "test-uid", decryptRequest); err != nil {
+			t.Fatalf("Decrypt() error = %v", err)
+		}
+		return true
+	})
+
+	// Out of the 8 calls to Decrypt:
+	// - 5 should be using the current local KEK
+	// - 1 out of the 3 should generate a decrypt call to the remote KMS as the local KEK not in cache
+	// - 2 out of the 3 should use the cached local KEK after 1st decrypt call to the remote KMS
+	assertCallCount(t, remoteKMS, localKEKService)
```

Comment:
I am still confused by this comment - should there be assertions for the `8`, `5`, `1` and `2` mentioned?

Reply:
The comment is mostly for the reader here. By asserting cache size is `1` and the # of calls to decrypt is `1` in addition to being to able successfully decrypt all encrypted data, we're ensuring the expected behavior.